### ISSUE TITLE
Global variables for shared state not working

### ISSUE THREAD

**Issue by fmckeogh** [fermyon/spin#354]
The following Rust module always returns `1`, as opposed to incrementing for each request.

```rust
use {
    anyhow::Result,
    once_cell::sync::Lazy,
    spin_sdk::{
        http::{Request, Response},
        http_component,
    },
    std::sync::Mutex,
};

static GLOBAL_DATA: Lazy<Mutex<u64>> = Lazy::new(|| Mutex::new(0));

#[http_component]
fn upload_video(req: Request) -> Result<Response> {
    println!("{:?}", req.headers());

    *(GLOBAL_DATA.lock().unwrap()) += 1;

    Ok(http::Response::builder()
        .status(418)
        .body(Some(format!("{:?}", GLOBAL_DATA.lock()).into()))?)
}
```

Is it correct that each request is served by its own instance of the WASM component? What would be the recommended correct way of implementing this program?

**Comment by radu-matei** [comment]
Hi, @fmckeogh!
You are correct, each request is currently served by a new Wasm instance (there are a number of reasons why this is currently preferred, request isolation being one of them).

This means that the components that can be written cannot keep state, which must be external.
The interfaces to access those external sources are currently developed — we have work in progress for a generic key/value store (#346), an object storage (#165), and a specific implementation for key/value for Redis (#328, #347).

Writing a Rust component that uses Redis to persist state:

https://github.com/fermyon/spin/blob/151d126ad02b8b346005edd2815751b9ddcc2908/examples/rust-outbound-redis/src/lib.rs#L16-L37

And one in Go (based on #347):

https://github.com/fermyon/spin/blob/28c86c469943702078674269aaac9065332c1e57/examples/tinygo-outbound-redis/main.go#L22-L43

Both examples assume you have an existing Redis instance, but our goal with the generic interfaces (both #346 and #165) is, for scenarios where you don't need a specific feature of say Redis, to use a generic interface and have an underlying implementation for it configured by an operator.
For example, if using the key/value store, if running locally in a dev environment, that could be a JSON file, but when running in production, a cloud instance of a key value storage.

TL; DR — state should be external to your Wasm component, and we are working on WebAssembly interfaces for connecting to external state providers to make it easy to write such applications.

**Comment by fmckeogh** [comment]
@radu-matei Thank you so much for your response, that's very helpful!
